pkg/render/prometheus_api.go

```diff
@@ -0,0 +1,321 @@
+	"github.com/tigera/operator/pkg/components"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	"github.com/tigera/operator/pkg/render/common/podsecuritypolicy"
+	"github.com/tigera/operator/pkg/render/common/secret"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	calicoNodePrometheusServiceName = "calico-node-prometheus"
+
+	tigeraPrometheusAPIName             = "tigera-prometheus-api"
+	prometheusEndpointUrlEnvVarName     = "PROMETHEUS_ENDPOINT_URL"
+	prometheusOperatedHttpServiceScheme = "http"
+	prometheusOperatedHttpServiceHost   = PrometheusHTTPAPIServiceName + ".tigera-prometheus"
+
+	prometheusServiceListenAddrEnvVarName = "LISTEN_ADDR"
+
+	tigeraPrometheusServiceHealthEndpoint = "/health"
+
+	// tigeraPrometheusAPIListenPort defines the port that tigera-prometheus-service
+	// pod listens to and the TargetPort of calico-node-prometheus service.
+	// If not set it defaults to port 9090. In the scenario that the cluster
+	// is hosted on EKS and using Calico as it's CNI, tigera-prometheus-service
+	// will be HostNetwoked. This is configurable in the configMap passed to TigeraPrometheusAPI
+	tigeraPrometheusAPIListenPortFieldName = "tigeraPrometheusAPIListenPort"
+)
+
+func TigeraPrometheusAPI(client client.Client, cr *operator.InstallationSpec, pullSecrets []*corev1.Secret) (Component, error) {
+
+	return &tigeraPrometheusAPIComponent{
+		client:       client,
+		installation: cr,
+		pullSecrets:  pullSecrets,
+	}, nil
+}
+
+type tigeraPrometheusAPIComponent struct {
+	client                 client.Client
+	installation           *operator.InstallationSpec
+	pullSecrets            []*corev1.Secret
+	prometheusServiceImage string
+}
+
+func (p *tigeraPrometheusAPIComponent) ResolveImages(is *operator.ImageSet) error {
+	reg := p.installation.Registry
+	path := p.installation.ImagePath
+	prefix := p.installation.ImagePrefix
+
+	prometheusServiceImage, err := components.GetReference(components.ComponentTigeraPrometheusService, reg, path, prefix, is)
+	if err != nil {
+		return err
+	}
+
+	p.prometheusServiceImage = prometheusServiceImage
+
+	return nil
+}
+
+// Objects returns the lists of objects in this component that should be created and/or deleted during
+// rendering.
+func (p *tigeraPrometheusAPIComponent) Objects() (objsToCreate, objsToDelete []client.Object) {
+	// tigera-prometheus-objects
+	namespacedObjects := []client.Object{}
+
+	// place pullsecrets secrets under tigera-prometheus
+	secrets := secret.CopyToNamespace(common.TigeraPrometheusNamespace, p.pullSecrets...)
+	namespacedObjects = append(namespacedObjects, secret.ToRuntimeObjects(secrets...)...)
+
+	tigeraPrometheusApiConfigMap, err := p.getConfigMap()
```

Comment:
Sorry I was not clear about this but getting this configmap should still be done as part of the monitor reconcile. It was only the creation of the configmap should be done by the render if it did not already exist (if nil was passed in for the configmap to the render).

Reply:
https://github.com/tigera/operator/pull/1379/commits/15a53feb00afb31e12a5c3523cc9ea60837b0c2e